Add Glyph Offset Option to User Configuration

aaron-williamson opened the pull request:
Love the idea of this terminal, great work so far! I really wanted this so I went ahead and added it myself :smile:

Let me know if there's anything I missed for this change, or tests you would like to accompany these changes.

jwilm commented on src/config.rs:
```diff
@@ -1095,6 +1132,10 @@ impl Default for Font {
                 x: 0.0,
                 y: 0.0
+            },
+            glyph_offset: GlyphOffset {
```
It would be a lot cleaner to `#[derive(Default)]` for `GlyphOffset` and use `Default::default()` here.

jwilm commented on src/config.rs:
```diff
@@ -1041,6 +1069,9 @@ pub struct Font {
     offset: FontOffset,
 
+    /// Glyph offset within character cell
+    glyph_offset: GlyphOffset,
```
This should include `#[serde(default)]`

jwilm commented on src/config.rs:
```diff
@@ -984,6 +984,34 @@ impl FontOffset {
+impl GlyphOffset {
+    /// Horizontal position
+    #[inline]
+    pub fn x(&self) -> f32 {
```
I think making the properties `pub` would be the easiest thing. These methods can just go away then.

jwilm commented on font/src/ft/mod.rs:
```diff
@@ -120,8 +124,8 @@ impl ::Rasterize for FreeTypeRasterizer {
             c: c,
-            top: glyph.bitmap_top(),
-            left: glyph.bitmap_left(),
+            top: glyph.bitmap_top() + self.glyph_offset_y,
```
I don't think this is the right place to apply the offset. The `font` package should be very general. The offsets are very Alacritty-specific.

A better spot might be where the glyph cache calls into the rasterizer.

aaron-williamson commented:
I think a good solution would be as you suggested, I could just make changes so that the fonts are correctly rendered the center by default. Then the option for user-configurable glyph offset would be unnecessary. Though some may still like it for things like aligning powerline fonts properly or similar.

aaron-williamson commented:
I've pushed a commit addressing your comments. Let me know if I missed anything, I wasn't totally sure where you meant for where the glyph cache calls the rasterizer, so let me know if I got that (or anything else) wrong.

Pending discussion I can also do some work to make the default centered, and possibly remove the glyph offset setting all together, I know https://github.com/jwilm/alacritty/issues/174#issuecomment-271386829 is in favor of default being centered with the option existing as well. Though I agree with you in that I can't think of any other terminal that has in-cell offsets as a setting, for example iTerm2's spacing settings (which already exist in Alacritty) may be sufficient.

koraa commented:
Hi, I just tested the branch and it definitely works as a fix for the powerline issues it was experiencing.
Your concerns about removing this feature in the future do apply of course!

jwilm commented:
Thanks for testing this @koraa. I've thought about this a bit more, and I have a few concerns.

- This offset is applied globally instead of per face
- With decent font metrics, this might not be needed
- If we attempt to center the glyphs when a _-y_ cell dimension is specified, this may not be needed.

Going to hold off on merging this PR until we can maybe investigate some alternative solutions. The fact that other terminal emulators don't need this makes me think that one of these other solutions (or combination thereof) would resolve this issue.

overdrivenpotato commented:
@jwilm I've posted this in the issue thread, but I'll post it here again. Other terminal emulators seem to have the exact same issue, and it doesn't seem to be properly configurable:

iTerm2
![iTerm2](http://i.imgur.com/IraehtX.png)
Terminal.app
![Terminal.app](http://i.imgur.com/G9th4uk.png)
[Font used is Source Code Pro for Powerline](https://github.com/powerline/fonts/tree/master/SourceCodePro)

iTerm2 gives you an option to change the height of the selected font as well as add a non-ascii font and adjust it's height separately. But that changes the height of all lines and still isn't perfectly centered:

![iTerm2 vertical spacing](http://i.imgur.com/RP1wmxl.png)

Alacritty with glyph_offset for comparison:

![alacritty with glyph_offset](http://i.imgur.com/lXTguim.png)

Visually, alacritty with the offset looks better than any terminal I've tested / MacVim

jwilm commented:
I'm convinced we should have this feature.

I want to think about how this fits into rendering/rasterization/config long term; let me get back to you tomorrow with some thoughts.

aaron-williamson commented:
Good to know you've made a decision! I'm more than happy to carry through with the implementation. I would love to hear where you think it should go. I had a thought maybe the offsets should be passed as arguments to the `get_glyph` function? Or is there somewhere that would be better for that?

A thought I've had as well is that even with this feature, it would make the most sense to have sane defaults for font offset. That is maybe center the glyph's by default instead of putting them in the bottom left. I did some looking my self but could use some help regarding where exactly to go about changing those defaults. My first thought was still in the `get_glyph` function but I wasn't able to find the font metrics needed for the center calculations (glyph width/height, cell width/height) from inside there. I might be missing something but could use some help regardless.

jwilm commented:
Thinking along these lines:

- The `get_glyph` api on `font::Rasterizer` shouldn't know anything about the offsets Alacritty applies. That API should be a platform agnostic wrapper on underlying font libraries and not manipulate their output.
- The `font::Rasterizer` glyph metrics need to be passed through to Alacritty. Provide the subset common to all platforms. Maybe provide a `font::ext::FreeType` and etc. traits if platform-specific stuff is needed. Some of this is longer term, but our use of metrics is currently bad.
- Offsets should be applied by the Alacritty renderer just outside the `font::Rasterizer::get_glyph` API before putting the glyph into the cache. Basically, the `x`, `y` offsets should be adjusted here.
- Each font face should specify its own offset
- Need to think about how this will affect fallback fonts

xyin96 commented:
Hey, I was wondering if this was still being worked on.

I went ahead and pushed a few commits on top of this as for decoupling the `Rasterizer` from the `GlyphCache` on my fork (see here: https://github.com/xyin96/alacritty/pull/3). Not exactly sure how to move forward with this feature.

aaron-williamson commented:
Sorry for the messy git history, I'll make sure to squash that at some point. Anyway, I've moved the glyph offset to the renderer and made the font unaware of it. I saw the work in @xyin96's fork and I'm not sure whether or not scaling the offset is really necessary. It should probably be done the same way the cell-size offset is done.

I've also improved the Linux metrics which removes the need for those wonky defaults and renders much nicer with defaults than before. I left the old metrics in there for sake of comparison (though I removed a mistake where it was calculating double the height)

I haven't, however, been able to find a way to center the font by default very easily. I agree that each face should give a default offset of its own so it looks a little nicer. With a little experimentation I found that the unicode characters for some reason aren't affected by the offset, but didn't have time to look into why. I would also like to improve how I implemented the offset in the `GlyphCache`, as storing the x and the y values separately is ugly and I should be able to just store a reference to/copy of the `GlyphCache` object from the config in a `glyph_cache` variable, instead of `glyph_cache_x` and `glyph_cache_y`

Lastly, I'm not really able to test these changes on OS X apart from what Travis CI does, so if someone wants to test these changes on OS X, please do! Any suggestions as to how I can develop/test the OS X application as well (without a Mac of my own) are welcome.

jwilm commented:
@aaron-williamson looks like you've done a bunch of good work here. I'm especially excited about improvements to our consumption of FreeType metrics. Busy week here, I'll make time to dig into this over the weekend.

aaron-williamson commented:
So I've got the tests passing and the glyph offset works on both Linux + macOS. The code can probably still use a little cleaning up. I've also run into a problem with unicode characters, which don't seem to be affected by the offest. For example the '❯' characters I use for my prompt, as seen in macOS (the unicode characters are lower than ASCII characters):
![image](https://cloud.githubusercontent.com/assets/9038848/23592427/f7fa1078-01bd-11e7-89e3-fa9bbc7b2ef2.png)

And in Linux:
![image](https://cloud.githubusercontent.com/assets/9038848/23592543/e7a6b756-01bf-11e7-8a24-0bb25b279746.png)

boxofrox commented:
@aaron-williamson I spent some time playing with this patch and found that not all glyphs are moving vertically.  Text uniformly obeys the `glyph_offset` settings, but symbols appear to be ignored.  I'm using 'DejaVu Sans Mono For PowerLine Nerd Font`.  Have you encountered this?

![image](https://cloud.githubusercontent.com/assets/636691/24616589/977e14a2-185f-11e7-8492-701a1240e62e.png)

aaron-williamson commented:
Yes @boxofrox I have noticed that issue, as I mentioned in my last comment. Though I haven't had time yet to dig in and figure out why exactly this is the case.

boxofrox commented:
@aaron-williamson apologies.  I failed to understand your previous comment.

aaron-williamson commented:
No worries, @boxofrox thanks for the extra pair of eyes. Good to know it's not just me getting that, and glad to see others using this. Hopefully I (or someone else, if they want to help) can figure this out 😄

aaron-williamson commented:
@boxofrox @jwilm I've found and fixed the issue with certain characters not getting the offset applied and also done some cleanup of my previous work. Let me know if there's anything you'd like me to change or if we can finally take this!

Example of fixed offset, macOS:
![image](https://cloud.githubusercontent.com/assets/9038848/25308040/b361d9e2-2769-11e7-81e6-d8cde29a59da.png)
Example of fixed offset, Ubuntu:
![image](https://cloud.githubusercontent.com/assets/9038848/25308144/03d421ee-276c-11e7-80a8-d8419c3cf93e.png)

boxofrox commented:
Very nice @aaron-williamson.  I finally got this merged into my local build and it looks great. :1st_place_medal: 
![image](https://cloud.githubusercontent.com/assets/636691/25393987/ff7afae6-29aa-11e7-9e6b-67db94f31010.png)

jwilm commented on font/src/ft/mod.rs:
```diff
@@ -54,23 +54,23 @@ impl ::Rasterize for FreeTypeRasterizer {
+        // A face should always have a size and size metrics
+        let size_metrics = match face.size_metrics() {
+            Some(metrics) => metrics,
+            None => panic!("There was an error loading font size metrics"),
```
The `metrics` method returns a `Result`. Let's take advantage of that instead of panicking.

jwilm commented on src/config.rs:
```diff
@@ -1189,6 +1189,31 @@ impl FontOffset {
+/// undesirable. This gives the user a way to shift where the glyphs are
+/// displayed in their cells.
+#[derive(Clone, Copy, Debug, Deserialize)]
+pub struct GlyphOffset {
```
The layout of this type is identical to `FontOffset`. Naming details aside, it would be nice to avoid that duplication. From a config file perspective (_alacritty.yml_), the user won't see the actual type, so not having a distinct type would just be an implementation detail.

Maybe we can just call it something like `Delta` instead of `FontOffset`/`GlyphOffset`. Then it can be used for the cell size change (currently the config `font.offset` object), *global* glyph offset as you've added here, and in the future, per-face glyph offset.

aperezdc commented:
I've tried this locally as well and it works real nice, and I am looking forward to see this merged :+1:

aaron-williamson commented:
@jwilm No worries about the delay, life happens, I haven't been able to work on it much myself!

I believe I've addressed your comments, and I've also squashed the commit history to be a little more reasonable. Let me know if there's anything else!

jwilm commented:
Looks good; thanks for the patch!

jmacdonald commented:
This works beautifully; thanks to everyone involved!

OJFord commented:
Was a `glyph_size` considered at any point here?

iTerm supports the equivalent, as it can then be used in situations like boxofox's image to align glyphs with line-height:
![](https://cloud.githubusercontent.com/assets/636691/25393987/ff7afae6-29aa-11e7-9e6b-67db94f31010.png)

(i.e. make those circles meet the top and bottom of the bar for a rounded effect)

Or just generally if the glyphs appear too small relative to text at the desired font size.

maximbaz commented:
Or perhaps `size` should be configurable per font, just like `glyph_offset` should be (not global as they are now).

boxofrox commented:
As I understand it, one glyph is a single rendering of a character/code point/ [[1]](https://glyphsapp.com/tutorials/unicode) and is determined by the `font.size` (and maybe the dpi, too).

\[1]: https://glyphsapp.com/tutorials/unicode

`offset` adds padding around a glyph.  In my screenshot (referenced by @OJFord), I added 4 pixels to each glyph to add more vertical space between the lines.  This doesn't stretch glyph, nor do I want it to.

If a cell is the box a glyph is drawn in on your terminal display, its height is `font.size + offset.y`.  Without `glyph_offset`, the glyph is drawn against the top of the cell, as seen in the first screenshots above.  With `glyph_offset`, we can position the glyph in a cell as desired (centered in the case of my screenshot).

> (i.e. make those circles meet the top and bottom of the bar for a rounded effect)

The "circles" don't meet the top and bottom of the "bar" because the bar is created by coloring an entire cell with a background color, while the circles are glyphs rendered from a patched font I use and thus restricted to the `font.size`.  If I want the circle glyphs to meet the top and bottom of the bar, then I can reset the `offset.y` to zero to remove the extra padding.

> Was a glyph_size considered at any point here?

No, I doubt this was considered.  Every symbol drawn in my screenshot was defined by one font with a size uniformly applied.  I don't see a consistent and intuitive way to discriminate text from text-sized symbols from line-height-sized symbols, unless fonts encode this information.

> Or perhaps size should be configurable per font, just like glyph_offset should be (not global as they are now).

@maximbaz, in what sense?

Remove `size` and add `font.normal.size`, `font.bold.size`, and `font.italic.size`?  I don't understand the utility of such a change.  Why would you want an inconsistent font size?

maximbaz commented:
I should have definitely been more clear: I'm thinking in a context of font fallback feature  https://github.com/jwilm/alacritty/issues/45.

Suppose I'd define the following config:

```
font:
  normal:
    - family: monospace
    - family: PowerlineSymbols
```

Now most of the characters are rendered with a monospace font, but only those that are missing (powerline symbols) are being rendered by the PowerlineSymbols font.

Now it may happen that the glyph positioning is different in monospace and PowerlineSymbols fonts, and that is where `glyph_offset` might come in handy. Today it is a global configuration, it affects all font families at once, but long-term it has to be configurable per font family, so that I can potentially align glyphs of PowerlineSymbols font without affecting monospace font (this was initially mentioned [here](https://github.com/jwilm/alacritty/issues/45#issuecomment-302812668)).

What I was suggesting above is that `size` should also be configurable per font family. It might happen that Powerline font has smaller glyphs in general, so I could maybe increase the font size only for Powerline font while preserving the size of monospace font - and that would again align the two fonts and make everything beautiful 🙂

jwilm commented:
One use case we've had suggested for per-font sizes is to make CJK text more readable.
